src/main/java/frc/robot/poseestimator/limelights/SmartLimelights.java

```diff
@@ -0,0 +1,114 @@
+package frc.robot.poseestimator.limelights;
+
+import edu.wpi.first.math.Pair;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Transform2d;
+import frc.utils.GBSubsystem;
+import org.littletonrobotics.junction.Logger;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Optional;
+
+
+public class SmartLimelights extends GBSubsystem {
+
+	private List<Limelight> limelights;
+	private LimelightsHardware limelightHardware;
+
+	public SmartLimelights(String[] limelightNames) {
+		super(VisionConstants.DEFAULT_CONFIG.Logpath());
+
+		this.limelightHardware = new LimelightsHardware(limelightNames, VisionConstants.DEFAULT_CONFIG.HardwareLogpath());
+		this.limelights = limelightHardware.getAllLimelights();
+	}
+
+	public List<Optional<Pair<Pose2d, Double>>> getAll2DEstimates() {
+		ArrayList<Optional<Pair<Pose2d, Double>>> estimates = new ArrayList<>();
+
+		for (Limelight limelight : limelights) {
+			if (!filterOutLimelight(limelight)) {
+				estimates.add(limelight.getUpdatedPose2DEstimation());
+			}
+		}
+
+		return estimates;
+	}
+
+	public boolean isLimelightedOutputInTolerance(Limelight limelight) {
+		Pose2d limelightPosition;
+		Transform2d transformDifference;
+		Rotation2d rotationDifference;
+
+		// TODO: placeholder for pubsubs, when it'll be added.
+		// ik we shouldn't have todos but it's notable enough when it throws compile errors
+		Pose2d currentPoseObservation = NetworkTables...;
+
+		limelightPosition = limelight.getUpdatedPose2DEstimation().get().getFirst();
+		transformDifference = limelightPosition.minus(currentPoseObservation);
+		rotationDifference = limelightPosition.getRotation().minus(currentPoseObservation.getRotation());
+
+		return transformDifference.getTranslation().getNorm() <= VisionConstants.DEFAULT_CONFIG.PositionNormTolerance()
+			&& rotationDifference.getDegrees() <= VisionConstants.DEFAULT_CONFIG.rotationTolerance().getDegrees();
+	}
+
+	public boolean isAprilTagInProperHeight(Limelight limelight) {
+		boolean aprilTagHeightConfidence = Math.abs(limelight.getAprilTagHeight() - VisionConstants.APRIL_TAG_HEIGHT_METERS)
+			< VisionConstants.APRIL_TAG_HEIGHT_TOLERANCE_METERS;
+		return aprilTagHeightConfidence;
+	}
+
+	public boolean hasTarget(Limelight limelight) {
+		return limelight.getUpdatedPose2DEstimation().isPresent();
+	}
+
+	public boolean filterOutLimelight(Limelight limelight) {
+		return !(hasTarget(limelight) && isAprilTagInProperHeight(limelight) && isLimelightedOutputInTolerance(limelight));
+	}
```

Comment:
not good for runtime

Reply:
wdym? why

Reply:
Calling again and again for get all estimations